File: fibonacci_algos/fibonacci_last_digit/fibonacci_last_digit.py
# Uses python3
import sys

def get_fibonacci_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % 10

def get_fibonacci_last_digit_fast(n):
    if n <= 1:
        return n

    fibonacci = []
    fibonacci.append(0)
    fibonacci.append(1)

    for i in range(2, n+1):
        fibonacci.insert(i, (fibonacci[i-1] + fibonacci[i-2])%10)

    return fibonacci[n]


if __name__ == '__main__':
    #input = sys.stdin.read()
    n = int(input())
    print(get_fibonacci_last_digit_fast(n))

File: fibonacci_algos/fibonacci_last_digit/FibonacciLastDigit.java
import java.util.*;

public class FibonacciLastDigit {
    private static int getFibonacciLastDigitNaive(int n) {
        if (n <= 1)
            return n;

        int previous = 0;
        int current  = 1;

        for (int i = 0; i < n - 1; ++i) {
            int tmp_previous = previous;
            previous = current;
            current = tmp_previous + current;
        }

        return current % 10;
    }

    private static int getFibonacciLastDigitFast(int n) {
        if (n <= 1)
            return n;

        int[] fibonacci = new int[n + 1];

        fibonacci[0] = 0;
        fibonacci[1]  = 1;

        for (int i = 2; i <= n ; ++i) 
            fibonacci[i] = (fibonacci[i - 1] + fibonacci[i - 2]) % 10;
        

        return fibonacci[n];
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int c = getFibonacciLastDigitFast(n);
        System.out.println(c);
    }
}

